pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,455 @@
+	plans, _, err := client.Plans.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list instance types / plans: %v", err)
+	}
+	// ensure our requested plan is in those plans
+	validPlanNames := planProp(plans, "Name")
+	if missingPlans := itemsNotInList(validPlanNames, []string{c.InstanceType}); len(missingPlans) > 0 {
+		return fmt.Errorf("unknown instance type / plan: %s, acceptable plans: %s", strings.Join(missingPlans, ","), strings.Join(validPlanNames, ","))
+	}
+
+	return nil
+}
+
+func (p *provider) Create(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData, userdata string) (instance.Instance, error) {
+	c, _, pc, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+
+	imageName, err := getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Invalid operating system specified %q, details = %v", pc.OperatingSystem, err),
+		}
+	}
+
+	serverCreateOpts := &packngo.DeviceCreateRequest{
+		Hostname:     machine.Spec.Name,
+		UserData:     userdata,
+		ProjectID:    c.ProjectID,
+		Facility:     c.Facilities,
+		BillingCycle: c.BillingCycle,
+		Plan:         c.InstanceType,
+		OS:           imageName,
+		Tags: []string{
+			generateTag(string(machine.UID)),
+		},
+	}
+
+	device, res, err := client.Devices.Create(serverCreateOpts)
+	if err != nil {
+		return nil, packetErrorToTerminalError(err, res, "failed to create server")
+	}
+
+	return &packetDevice{device: device}, nil
+}
+
+func (p *provider) Cleanup(machine *v1alpha1.Machine, _ *cloud.MachineCreateDeleteData) (bool, error) {
+	instance, err := p.Get(machine)
+	if err != nil {
+		if err == cloudprovidererrors.ErrInstanceNotFound {
+			return true, nil
+		}
+		return false, err
+	}
+
+	c, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return false, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+
+	res, err := client.Devices.Delete(instance.(*packetDevice).device.ID)
+	if err != nil {
+		return false, packetErrorToTerminalError(err, res, "failed to delete the server")
+	}
+
+	return false, nil
+}
+
+func (p *provider) AddDefaults(spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	return spec, nil
+}
+
+func (p *provider) Get(machine *v1alpha1.Machine) (instance.Instance, error) {
+	device, _, err := p.getPacketDevice(machine)
+	if err != nil {
+		return nil, err
+	}
+	if device != nil {
+		return &packetDevice{device: device}, nil
+	}
+
+	return nil, cloudprovidererrors.ErrInstanceNotFound
```

Comment:
If the API can not find the instance it does return an empty error and a nil `device`...?

Reply:
Oh, did I get it backwards? Oops. Fixing.

Reply:
No, wait, that makes sense. It returns `nil` for the device and `cloudprovidererrors.ErrInstanceNotFound` for the error. I took this from all of the other cloud providers, see [here](https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/aws/provider.go#L602), [here](https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/azure/provider.go#L558), [here](https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/hetzner/provider.go#L288), etc.

Reply:
Okay it actually does work, because `getDeviceByTag` returns a `nil` error if it can't find the device. I personally find that subpar, because this means all users of `p.getPacketDevice` must check if the returned `device` is non-nil (Which they currently don't do). What about making `getDeviceByTag` return an error if it cant find the device?

Reply:
I am having a hard time following your logic. I will try to work it through here.

1. `getDeviceByTag()` (internal only): gets the device that matches the tag, or `nil` if none is found. Returns a non-nil `error` if there is a real error. 
2. `p.getPacketDevice()` (internal only): calls `getDeviceByTag()`. It really just wraps `getDeviceByTag()` while providing some additional convenience (I probably could subsume `getDeviceByTag()` into it, since it no longer is used elsewhere, but I like small, easy-to-understand functions). It returns the device if found, `nil` if not; `error` only if there was an error.
3. The internal convenience function `p.getPacketDevice()` is consumed by `p.Get()` and `p.MigrateUID()`, which check that... 

Ah, I see. You are saying that the API from `p.getPacketDevice()` is "return `nil` if no device found", but not everyone is respecting that, particularly `p.MigrateUID()`.

OK, that is a bug in my implementation of `p.MigrateUID()`. I will fix that.

I would like to leave `p.getPacketDevice()` as is, as I tend to view 3 distinct use cases:

* found the device: `device` set; `err = nil`
* had a comms or auth error: `device = nil`; `err` set
* no error but no matching device: `device = nil`; `err = nil`.

Reply:
I took inspiration from the Hetzner and AWS code: `glog` a message and return `nil`.

Reply:
Updated, `lint` and `build` all look fine.

Reply:
Agree except for `no error but no matching device: device = nil; err = nil.`. That is counterintuitive, if I get no error I expect the `device` to be `non-nil`. Also it makes for redundant nil-checks everywhere `p.getPacketDevice` is used.

Where in the Hetzner or AWS code do you see a `nil` error return when no instance was found?

Reply:
> Agree except for no error but no matching device: device = nil; err = nil.. That is counterintuitive, if I get no error I expect the device to be non-nil

To each their own. Since this is internal to the provider, we can agree to disagree. 

> Where in the Hetzner or AWS code do you see a nil error return when no instance was found?

In `MigrateUID()` [AWS here](https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/aws/provider.go#L646) and [Hetzner here](https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/hetzner/provider.go#L310)

Reply:
`MigrateUID` is a special case and only returns a nil error if `p.Get` returned `err == cloudprovidererrors.ErrInstanceNotFound`.

This is really prone to result in bugs in the future. Why would I assume that if I call `p.getPacketDevice()` and `err == nil` instance can also be nil and has to be checked separately?

Reply:
Like what would you think of a rest api that returned for a `GET` call a http/200 and no body?

Reply:
> Why would I assume that if I call p.getPacketDevice() and err == nil device can also be nil and has to be checked separately?
 
That is pretty common, actually. 

> Like what would you think of a rest api that returned for a GET call a http/200 and no body?
 
It depends. `GET /items/1234` _should_ return a `404` for not found, and a `200` for found. But `GET /items` or a search like `GET /items?name=jim` should return `404` only if the path `GET /items` does not exist. If none matches `name=jim`, then it should return a `200` with an empty set. That has been a longstanding debate among REST implementers/builders for about as long as Fielding's paper has been out (almost 30 years), but I always came down on that side.

I think, though, that we are spending too much time on this. It is a purely internal private function. It works, is clearly defined. Let's just leave it?